src/pocketmine/inventory/BaseUIInventory.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+*/
+
+declare(strict_types=1);
+
+namespace pocketmine\inventory;
+
+use pocketmine\item\Item;
+use pocketmine\item\ItemFactory;
+use pocketmine\Player;
+
+class BaseUIInventory extends BaseInventory{
```

Comment:
The introduction of this class pushes network internals off onto plugin developers for the sake of catering for one change: the addressing of fake inventories is now mapped to window ID `124` and using magic slot numbers, instead of negative window IDs and normal slot numbers.

This is an overcomplicated solution for such a problem which would be better solved at a network-local level with a more isolated hack that doesn't touch the API.